ddl/ddl_api.go

```diff
@@ -1365,41 +1365,41 @@ func buildViewInfoWithTableColumns(ctx sessionctx.Context, s *ast.CreateViewStmt
 	return viewInfo, tableColumns
 }
 
-func checkPartitionByHash(pi *model.PartitionInfo) error {
+func checkPartitionByHash(ctx sessionctx.Context, pi *model.PartitionInfo, s *ast.CreateTableStmt, tbInfo *model.TableInfo) error {
 	if err := checkAddPartitionTooManyPartitions(pi.Num); err != nil {
-		return errors.Trace(err)
+		return err
 	}
-	if err := checkNoHashPartitions(pi.Num); err != nil {
-		return errors.Trace(err)
+	if err := checkNoHashPartitions(ctx, pi.Num); err != nil {
+		return err
 	}
-	return nil
+	if err := checkPartitionFuncValid(ctx, tbInfo, s.Partition.Expr); err != nil {
+		return err
+	}
+	return checkPartitionFuncType(ctx, s, nil, tbInfo)
```

Comment:
```suggestion
	return checkPartitionFuncType(ctx, s, tbInfo.Columns, tbInfo)
```
or remove cols paraments, it is redundant?